### PR TITLE
Add Next Up indicator and some clean up

### PR DESCRIPTION
**Pull request by bigredwill**
I got a little carried away. Let me know if I should remove anything or create multiple PRs

New functionality:
- Next Up view. This is on the right side below the rotate button.

Changed:
- added .gitignore
- added grunt-cli to package.json along with `npm run build:watch` script helper.
- center canvases with html/css
- utilize window.requestAnimationFrame for game loop
  - Probably overkill for this use case, but I wanted to try it out
- Improved pause functionality
  - fixes issue with resuming pause causing new piece to spawn
  - Adds 'Pause' text and transparent overlay to indicate pause state
 - Modify keydown listeners so they are grouped by GAME_STATE
   - Space bar can start game and skip tutorial  
 

Screen Shot 2021-07-01 at 11 09 47 PM

### CODE REVIEW

**Comment by bigredwill** [comment]
Only todo is centering the up next tetris piece. My brain is too tired for that math right now

File: index.html [hunk 1]
```diff
@@ -25,33 +25,51 @@
   </head>
 
   <body>
-    <canvas id="background-canvas"></canvas>
```
Comment (bigredwill) [inline index.html:28]: moved canvases to children of `.layout-center` on line 43

File: index.html [hunk 2]
```diff
@@ -25,33 +25,51 @@
...
+        </div>
+    
+        <div class="divbutton centering-flex-wrapper" id="button-rotate">
+          <div class="divbutton-body unselectable-text">ROTATE<br/>(&#x2191;&#xfe0e;)</div>
+        </div>
 
-    <div class="divbutton centering-flex-wrapper" id="button-rotate">
-      <div class="divbutton-body unselectable-text">ROTATE<br/>(&#x2191;&#xfe0e;)</div>
+        <div class="divbutton centering-flex-wrapper" id="button-rotate">
+          <canvas id="info-canvas"></canvas>
```
Comment (bigredwill) [inline index.html:58]: adds new `info-canvas` for showing up next piece

File: src/board-painter.js [hunk 3]
```diff
@@ -35,59 +36,65 @@ const DEFAULT_COLOR = '#C39953'; // Aztec Gold
...
     this.fgctx = this.foregroundCanvas.getContext('2d');
+    this.infoctx = this.infoCanvas.getContext('2d');
 
     this.initialize();
   }
 
   initialize() {
-    this.resizeCanvasToFillWindow(CANVAS.FOREGROUND);
-    this.resizeCanvasToFillWindow(CANVAS.BACKGROUND);
+    this.resizeCanvasToFillParent(CANVAS.FOREGROUND);
```
Comment (bigredwill) [inline src/board-painter.js:55]: canvases take up entirety of containing element width/height

File: src/board-painter.js [hunk 4]
```diff
@@ -35,59 +36,65 @@ const DEFAULT_COLOR = '#C39953'; // Aztec Gold
...
+    this.resizeCanvasToFillParent(CANVAS.BACKGROUND);
+    this.resizeCanvasToFillParent(CANVAS.INFO);
+
+    this.localizeBoard();
```
Comment (bigredwill) [inline src/board-painter.js:59]: since board is only used in `CANVAS.FOREGROUND`, removed that argument

File: src/board-painter.js [hunk 5]
```diff
@@ -97,6 +104,34 @@ class BoardPainter {
     };
   }
 
+  /**
+   * todo: center entire piece in info view
+   */
+  infoBoardToWorld(pixel) {
```
Comment (bigredwill) [inline src/board-painter.js:110]: this needs work as the pieces are not centered within the canvas. This function does not know about the entire piece, just individual pixels. 

We could calculate the bounding box for the piece, then make the x and y values relative to the top-left of the bounding box

File: src/game.js [hunk 6]
```diff
@@ -10,6 +10,8 @@ const Tutorial = require('./tutorial');
 const GAME_STATE = {
   RUNNING: 'RUNNING',
   PAUSED: 'PAUSED',
+  TUTORIAL: 'TUTORIAL',
+  MENU: 'MENU',
```
Comment (bigredwill) [inline src/game.js:14]: added `TUTORIAL` and `MENU` game states to make it easier to separate out event actions per state

File: src/game.js [hunk 7]
```diff
@@ -136,31 +144,40 @@ class Game {
...
 
   handleKeypress(event) {
-    if (event.code === 'ArrowLeft') {
-      this.tryToMoveLeft();
-    }
 
-    if (event.code === 'ArrowRight') {
-      this.tryToMoveRight();
-    }
+    if (this.state === GAME_STATE.RUNNING) {
```
Comment (bigredwill) [inline src/game.js:149]: these happen in the RUNNING state only

File: src/game.js [hunk 8]
```diff
@@ -136,31 +144,40 @@ class Game {
...
+      if (event.code === 'Space') {
+        this.tryToDrop();
+      }
     }
 
-    if (event.code === 'Space') {
-      this.tryToDrop();
+    if (this.state === GAME_STATE.MENU) {
+      if (event.code === 'Space') {
+        this.tryToStart();
```
Comment (bigredwill) [inline src/game.js:173]: Allows space bar to start the game

File: src/game.js [hunk 9]
```diff
@@ -136,31 +144,40 @@ class Game {
...
+      }
     }
 
     if (event.code === 'KeyP') {
       if (this.state === GAME_STATE.PAUSED) {
-        this.startGame();
-      } else {
-        this.pause();
+        this.runGame();
```
Comment (bigredwill) [inline src/game.js:179]: pauseGame is now paired with runGame - this just pauses/unpauses the game loop

File: src/game.js [hunk 10]
```diff
@@ -239,24 +256,48 @@ class Game {
...
-    }
+  gameLoop(tFrame) {
+    this.stopGameLoop = window.requestAnimationFrame(this.gameLoop.bind(this));
 
-    this.refreshProcessId = setInterval(() => {
+
+    const nextStepTime = this.lastStepTime + REFRESH_RATE_MS;
+    let numTicks = 0;
+
+    if (tFrame > nextStepTime) {
```
Comment (bigredwill) [inline src/game.js:268]: we keep track of the `lastStepTime` so that we can call `step()` after the correct  amount of time has passed (`nextStepTime`)

Comment (bigredwill) [reply src/game.js:268]: `window.requestAnimationFrame` is a more performant way of doing constantly looping tasks https://developer.mozilla.org/en-US/docs/Web/API/window/requestAnimationFrame

Comment (bigredwill) [reply src/game.js:268]: Additionally - this opens up animations on the canvas using linear interpolation (lerp) and some easing functions https://gist.github.com/greggman/7ec38b9d19546f15d610ddb82610646a

File: style.css [hunk 11]
```diff
@@ -46,69 +46,76 @@ canvas {
...
 }
 
 .divbutton-body {
     font-family: 'VT323', monospace;
     font-size: 30px;
     color: rgb(100, 100, 100);
     text-align: center;
+    width: 100%;
+    height: 100%;
+    display: flex;
```
Comment (bigredwill) [inline style.css:66]: flex centering of contents vertically and horizontally

**Comment by kevinmcvey** [comment]
Heck yeah this is so cool, thanks for all the updates <3

**Pull request merged**